File: code/backend/php/src/Core/Service/Story/ContentQueryService.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Service\Story;

use Games\Paths\Core\Domain\Story\CardInfo;
use Games\Paths\Core\Domain\Story\CreatorInfo;
use Games\Paths\Core\Domain\Story\TextInfo;
use Games\Paths\Core\Port\Story\ContentQueryPort;
use Games\Paths\Core\Port\Story\StoryReadPort;

class ContentQueryService implements ContentQueryPort
{
    private StoryReadPort $readPort;

    public function __construct(StoryReadPort $readPort)
    {
        $this->readPort = $readPort;
    }

    public function getCardByStoryAndCardUuid(string $storyUuid, string $cardUuid, string $lang): ?CardInfo
    {
        if (empty(trim($storyUuid)) || empty(trim($cardUuid))) {
            return null;
        }

        $story = $this->readPort->findStoryByUuid($storyUuid);
        if (!$story) {
            return null;
        }

        $storyId = (int) $story['id'];
        $card = $this->readPort->findCardByStoryIdAndUuid($storyId, $cardUuid);
        if (!$card) {
            return null;
        }

        $title = $this->resolveText($storyId, isset($card['id_text_title']) ? (int)$card['id_text_title'] : null, $lang);
        $description = $this->resolveText($storyId, isset($card['id_text_description']) ? (int)$card['id_text_description'] : null, $lang);
        $copyrightText = $this->resolveText($storyId, isset($card['id_text_copyright']) ? (int)$card['id_text_copyright'] : null, $lang);
        $creator = $this->resolveCreator($storyId, isset($card['id_creator']) ? (int)$card['id_creator'] : null, $lang);

        return new CardInfo(
            $card['uuid'] ?? '',
            $card['image_url'] ?? null,
            $card['alternative_image'] ?? null,
            $card['awesome_icon'] ?? null,
            $card['style_main'] ?? null,
            $card['style_detail'] ?? null,
            $title,
            $description,
            $copyrightText,
            $card['link_copyright'] ?? null,
            $creator
        );
    }

    public function getTextByStoryAndIdText(string $storyUuid, int $idText, string $lang): ?TextInfo
    {
        if (empty(trim($storyUuid))) {
            return null;
        }

        $story = $this->readPort->findStoryByUuid($storyUuid);
        if (!$story) {
            return null;
        }

        $storyId = (int) $story['id'];
        $effectiveLang = (!empty(trim($lang))) ? $lang : 'en';

        $text = $this->readPort->findTextByStoryIdTextAndLang($storyId, $idText, $effectiveLang);
        $resolvedLang = $effectiveLang;

        if (!$text && $effectiveLang !== 'en') {
            $text = $this->readPort->findTextByStoryIdTextAndLang($storyId, $idText, 'en');
            if ($text) {
                $resolvedLang = 'en';
            }
        }

        if (!$text) {
            return null;
        }

        $copyrightText = $this->resolveText($storyId, isset($text['id_text_copyright']) ? (int)$text['id_text_copyright'] : null, $effectiveLang);
        $creator = $this->resolveCreator($storyId, isset($text['id_creator']) ? (int)$text['id_creator'] : null, $effectiveLang);

        return new TextInfo(
            (int) $text['id_text'],
            $effectiveLang,
            $resolvedLang,
            $text['short_text'] ?? null,
            $text['long_text'] ?? null,
            $copyrightText,
            $text['link_copyright'] ?? null,
            $creator
        );
    }

    public function getCreatorByStoryAndCreatorUuid(string $storyUuid, string $creatorUuid, string $lang): ?CreatorInfo
    {
        if (empty(trim($storyUuid)) || empty(trim($creatorUuid))) {
            return null;
        }

        $story = $this->readPort->findStoryByUuid($storyUuid);
        if (!$story) {
            return null;
        }

        $storyId = (int) $story['id'];
        $creator = $this->readPort->findCreatorByStoryIdAndUuid($storyId, $creatorUuid);
        if (!$creator) {
            return null;
        }

        $name = $this->resolveText($storyId, isset($creator['id_text']) ? (int)$creator['id_text'] : null, $lang);

        return new CreatorInfo(
            $creator['uuid'] ?? '',
            $name,
            $creator['link'] ?? null,
            $creator['url'] ?? null,
            $creator['url_image'] ?? null,
            $creator['url_emote'] ?? null,
            $creator['url_instagram'] ?? null
        );
    }

    // === Private helpers ===

    public function resolveText(int $storyId, ?int $idText, string $lang): ?string
    {
        if ($idText === null) {
            return null;
        }
        $effectiveLang = (!empty(trim($lang))) ? $lang : 'en';

        $text = $this->readPort->findTextByStoryIdTextAndLang($storyId, $idText, $effectiveLang);
        if ($text) {
            return $text['short_text'] ?? null;
        }

        if ($effectiveLang !== 'en') {
            $fallback = $this->readPort->findTextByStoryIdTextAndLang($storyId, $idText, 'en');
            if ($fallback) {
                return $fallback['short_text'] ?? null;
            }
        }

        return null;
    }

    public function resolveCreator(int $storyId, ?int $idCreator, string $lang): ?CreatorInfo
    {
        if ($idCreator === null) {
            return null;
        }

        $creators = $this->readPort->findCreatorsForStory($storyId);
        foreach ($creators as $c) {
            if (isset($c['id']) && (int)$c['id'] === $idCreator) {
                $name = $this->resolveText($storyId, isset($c['id_text']) ? (int)$c['id_text'] : null, $lang);
                return new CreatorInfo(
                    $c['uuid'] ?? '',
                    $name,
                    $c['link'] ?? null,
                    $c['url'] ?? null,
                    $c['url_image'] ?? null,
                    $c['url_emote'] ?? null,
                    $c['url_instagram'] ?? null
                );
            }
        }

        return null;
    }
}

File: code/backend/python/app/core/services/story/content_query_service.py
from typing import Optional
from app.core.models.story.card_info import CardInfo
from app.core.models.story.text_info import TextInfo
from app.core.models.story.creator_info import CreatorInfo
from app.core.ports.story.content_query_port import ContentQueryPort
from app.core.ports.story.story_read_port import StoryReadPort


class ContentQueryService(ContentQueryPort):
    def __init__(self, read_port: StoryReadPort):
        self.read_port = read_port

    def get_card_by_story_and_card_uuid(self, story_uuid: str, card_uuid: str, lang: str) -> Optional[CardInfo]:
        if not story_uuid or not story_uuid.strip():
            return None
        if not card_uuid or not card_uuid.strip():
            return None

        story = self.read_port.find_story_by_uuid(story_uuid)
        if not story:
            return None

        story_id = story.get("id")
        card = self.read_port.find_card_by_story_id_and_uuid(story_id, card_uuid)
        if not card:
            return None

        title = self._resolve_text(story_id, card.get("id_text_title"), lang)
        description = self._resolve_text(story_id, card.get("id_text_description"), lang)
        copyright_text = self._resolve_text(story_id, card.get("id_text_copyright"), lang)
        creator = self._resolve_creator(story_id, card.get("id_creator"), lang)

        return CardInfo(
            uuid=card.get("uuid"),
            imageUrl=card.get("image_url"),
            alternativeImage=card.get("alternative_image"),
            awesomeIcon=card.get("awesome_icon"),
            styleMain=card.get("style_main"),
            styleDetail=card.get("style_detail"),
            title=title,
            description=description,
            copyrightText=copyright_text,
            linkCopyright=card.get("link_copyright"),
            creator=creator
        )

    def get_text_by_story_and_id_text(self, story_uuid: str, id_text: int, lang: str) -> Optional[TextInfo]:
        if not story_uuid or not story_uuid.strip():
            return None

        story = self.read_port.find_story_by_uuid(story_uuid)
        if not story:
            return None

        story_id = story.get("id")
        effective_lang = lang if lang and lang.strip() else "en"

        text = self.read_port.find_text_by_story_id_text_and_lang(story_id, id_text, effective_lang)
        resolved_lang = effective_lang

        if not text and effective_lang != "en":
            text = self.read_port.find_text_by_story_id_text_and_lang(story_id, id_text, "en")
            if text:
                resolved_lang = "en"

        if not text:
            return None

        copyright_text = self._resolve_text(story_id, text.get("id_text_copyright"), effective_lang)
        creator = self._resolve_creator(story_id, text.get("id_creator"), effective_lang)

        return TextInfo(
            idText=text.get("id_text"),
            lang=effective_lang,
            resolvedLang=resolved_lang,
            shortText=text.get("short_text"),
            longText=text.get("long_text"),
            copyrightText=copyright_text,
            linkCopyright=text.get("link_copyright"),
            creator=creator
        )

    def get_creator_by_story_and_creator_uuid(self, story_uuid: str, creator_uuid: str, lang: str) -> Optional[CreatorInfo]:
        if not story_uuid or not story_uuid.strip():
            return None
        if not creator_uuid or not creator_uuid.strip():
            return None

        story = self.read_port.find_story_by_uuid(story_uuid)
        if not story:
            return None

        story_id = story.get("id")
        creator = self.read_port.find_creator_by_story_id_and_uuid(story_id, creator_uuid)
        if not creator:
            return None

        name = self._resolve_text(story_id, creator.get("id_text"), lang)

        return CreatorInfo(
            uuid=creator.get("uuid"),
            name=name,
            link=creator.get("link"),
            url=creator.get("url"),
            urlImage=creator.get("url_image"),
            urlEmote=creator.get("url_emote"),
            urlInstagram=creator.get("url_instagram")
        )

    # === Private helpers ===

    def _resolve_text(self, story_id: int, id_text: Optional[int], lang: str) -> Optional[str]:
        if id_text is None:
            return None
        effective_lang = lang if lang and lang.strip() else "en"

        text = self.read_port.find_text_by_story_id_text_and_lang(story_id, id_text, effective_lang)
        if text:
            return text.get("short_text")

        if effective_lang != "en":
            fallback = self.read_port.find_text_by_story_id_text_and_lang(story_id, id_text, "en")
            if fallback:
                return fallback.get("short_text")

        return None

    def _resolve_creator(self, story_id: int, id_creator: Optional[int], lang: str) -> Optional[CreatorInfo]:
        if id_creator is None:
            return None

        creators = self.read_port.find_creators_for_story(story_id)
        for c in creators:
            if c.get("id") is not None and int(c["id"]) == id_creator:
                name = self._resolve_text(story_id, c.get("id_text"), lang)
                return CreatorInfo(
                    uuid=c.get("uuid"),
                    name=name,
                    link=c.get("link"),
                    url=c.get("url"),
                    urlImage=c.get("url_image"),
                    urlEmote=c.get("url_emote"),
                    urlInstagram=c.get("url_instagram")
                )
        return None
